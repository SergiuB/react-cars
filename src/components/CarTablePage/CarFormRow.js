import React from 'react';
import PropTypes from 'prop-types';

import CarForm from '../CarForm';

const CarFormRow = ({ colSpan, ...rest }) => (
  <tr>
    <td colSpan={colSpan}>
      <CarForm {...rest} />
    </td>
  </tr>
);

CarFormRow.propTypes = {
  colSpan: PropTypes.number.isRequired,
};

CarFormRow.defaultProps = {
  colSpan: 1,
};

export default CarFormRow;
