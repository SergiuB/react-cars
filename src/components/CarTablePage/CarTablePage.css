.car-table-page .id-col {
  width: 40px;
}

.car-table-page .large-col {
  width: 120px;
}

.car-table-page .small-col {
  width: 60px;
}