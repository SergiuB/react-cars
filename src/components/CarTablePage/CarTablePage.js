import React from 'react';
import PropTypes from 'prop-types';

import CarRow from '../CarRow';
import CarFormRow from './CarFormRow';
import carShape from '../../carShape';

const getCarRows = (cars, editingCarId, onRowClick) => cars.map(car => (
  <CarRow
    key={car.id}
    car={car}
    editEnabled={car.id === editingCarId}
    onClick={() => onRowClick(car.id)}
  />
));

const findCarIndexById = (cars, carId) => cars.findIndex(car => car.id === carId);

const CarTablePage = ({
  cars,
  editingCarId,
  onSaveCarChanges,
  onCancelCarChanges,
  onRowClick,
}) => {
  const allRows = getCarRows(cars, editingCarId, onRowClick);

  if (editingCarId !== -1) {
    const editingCarIndex = findCarIndexById(cars, editingCarId);
    if (editingCarIndex !== -1) {
      // insert a CarFormRow for the car to be edited
      const car = cars[editingCarIndex];
      allRows.splice(editingCarIndex + 1, 0, (
        <CarFormRow
          key={`${car.id}-form`}
          car={car}
          onSubmit={onSaveCarChanges}
          onCancel={onCancelCarChanges}
        />
      ));
    }
  }

  return (
    <table>
      <thead>
        <tr>
          <th>Id</th>
          <th>Name</th>
          <th>acceleration</th>
        </tr>
      </thead>
      <tbody>
        {allRows}
      </tbody>
    </table>
  );
};

CarTablePage.propTypes = {
  cars: PropTypes.arrayOf(carShape).isRequired,
  editingCarId: PropTypes.number.isRequired,
  onSaveCarChanges: PropTypes.func.isRequired,
  onCancelCarChanges: PropTypes.func.isRequired,
  onRowClick: PropTypes.func.isRequired,
};


CarTablePage.defaultProps = {
  editingCarId: -1,
  onSaveCarChanges: () => {},
  onCancelCarChanges: () => {},
  onRowClick: () => {},
};

export default CarTablePage;
