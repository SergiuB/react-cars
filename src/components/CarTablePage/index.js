import CarTablePage from './CarTablePage';

export default CarTablePage;
