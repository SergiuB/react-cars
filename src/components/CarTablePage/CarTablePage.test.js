import React from 'react';
import { shallow } from 'enzyme';

import CarTablePage from './CarTablePage';

const cars = [{
  id: 1,
  name: 'Audi',
  acceleration: 12,
}, {
  id: 2,
  name: 'VW',
  acceleration: 18,
}];

describe('CarTablePage', function () {
  it('renders correctly with no edited car', function () {
    const wrapper = shallow(
      <CarTablePage
        cars={cars}
      />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('CarRow')).toHaveLength(2);
    expect(wrapper.find('CarFormRow')).toHaveLength(0);
  });

  it('renders correctly with first car in edit mode', function () {
    const wrapper = shallow(
      <CarTablePage
        cars={cars}
        editingCarId={1}
      />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('CarRow')).toHaveLength(2);
    expect(wrapper.find('CarFormRow').prop('car').id).toEqual(1);
  });

  it('renders correctly with last car in edit mode', function () {
    const wrapper = shallow(
      <CarTablePage
        cars={cars}
        editingCarId={2}
      />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('CarRow')).toHaveLength(2);
    expect(wrapper.find('CarFormRow').prop('car').id).toEqual(2);
  });

  it('renders correctly with invalid editingCarId', function () {
    const wrapper = shallow(
      <CarTablePage
        cars={cars}
        editingCarId={100}
      />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('CarRow')).toHaveLength(2);
    expect(wrapper.find('CarFormRow')).toHaveLength(0);
  });

  it('fires onRowClick with the car id as argument when row is clicked', function () {
    const onRowClick = jest.fn();
    const wrapper = shallow(
      <CarTablePage
        cars={cars}
        onRowClick={onRowClick}
      />
    );
    // click on the 2nd row on the first page
    wrapper.find('CarRow').at(1).prop('onClick')();
    expect(onRowClick).toBeCalledWith(2);
  });

  it('fires onSaveCarChanges with updated car if inner form is submitted', function () {
    const updatedCar = { id: 2, name: 'Faster VW', acceleration: 5 };
    const onSaveCarChanges = jest.fn();
    const wrapper = shallow(
      <CarTablePage
        cars={cars}
        editingCarId={2}
        onSaveCarChanges={onSaveCarChanges}
      />
    );
    wrapper.find('CarFormRow').prop('onSubmit')(updatedCar);
    expect(onSaveCarChanges).toBeCalledWith(updatedCar);
  });

  it('fires onCancelCarChanges if inner form is cancelled', function () {
    const onCancelCarChanges = jest.fn();
    const wrapper = shallow(
      <CarTablePage
        cars={cars}
        editingCarId={2}
        onCancelCarChanges={onCancelCarChanges}
      />
    );
    wrapper.find('CarFormRow').prop('onCancel')();
    expect(onCancelCarChanges).toBeCalled();
  });
});
