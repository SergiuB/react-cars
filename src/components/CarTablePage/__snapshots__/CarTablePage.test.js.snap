// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarTablePage renders correctly with first car in edit mode 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <table>
    <thead>
        <tr>
            <th>
                Id
            </th>
            <th>
                Name
            </th>
            <th>
                acceleration
            </th>
        </tr>
    </thead>
    <tbody>
        <CarRow
            car={
                Object {
                    "acceleration": 12,
                    "id": 1,
                    "name": "Audi",
                  }
            }
            editEnabled={true}
            onClick={[Function]}
        />
        <CarFormRow
            car={
                Object {
                    "acceleration": 12,
                    "id": 1,
                    "name": "Audi",
                  }
            }
            onCancel={[Function]}
            onSubmit={[Function]}
        />
        <CarRow
            car={
                Object {
                    "acceleration": 18,
                    "id": 2,
                    "name": "VW",
                  }
            }
            editEnabled={false}
            onClick={[Function]}
        />
    </tbody>
</table>,
  "nodes": Array [
    <table>
      <thead>
            <tr>
                  <th>
                        Id
                  </th>
                  <th>
                        Name
                  </th>
                  <th>
                        acceleration
                  </th>
            </tr>
      </thead>
      <tbody>
            <CarRow
                  car={
                        Object {
                              "acceleration": 12,
                              "id": 1,
                              "name": "Audi",
                            }
                  }
                  editEnabled={true}
                  onClick={[Function]}
            />
            <CarFormRow
                  car={
                        Object {
                              "acceleration": 12,
                              "id": 1,
                              "name": "Audi",
                            }
                  }
                  onCancel={[Function]}
                  onSubmit={[Function]}
            />
            <CarRow
                  car={
                        Object {
                              "acceleration": 18,
                              "id": 2,
                              "name": "VW",
                            }
                  }
                  editEnabled={false}
                  onClick={[Function]}
            />
      </tbody>
</table>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <CarTablePage
        cars={
                Array [
                        Object {
                          "acceleration": 12,
                          "id": 1,
                          "name": "Audi",
                        },
                        Object {
                          "acceleration": 18,
                          "id": 2,
                          "name": "VW",
                        },
                      ]
        }
        editingCarId={1}
        onCancelCarChanges={[Function]}
        onRowClick={[Function]}
        onSaveCarChanges={[Function]}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "cars": Array [
            Object {
              "acceleration": 12,
              "id": 1,
              "name": "Audi",
            },
            Object {
              "acceleration": 18,
              "id": 2,
              "name": "VW",
            },
          ],
          "editingCarId": 1,
          "onCancelCarChanges": [Function],
          "onRowClick": [Function],
          "onSaveCarChanges": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <table>
          <thead>
                    <tr>
                              <th>
                                        Id
                              </th>
                              <th>
                                        Name
                              </th>
                              <th>
                                        acceleration
                              </th>
                    </tr>
          </thead>
          <tbody>
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 12,
                                                  "id": 1,
                                                  "name": "Audi",
                                                }
                              }
                              editEnabled={true}
                              onClick={[Function]}
                    />
                    <CarFormRow
                              car={
                                        Object {
                                                  "acceleration": 12,
                                                  "id": 1,
                                                  "name": "Audi",
                                                }
                              }
                              onCancel={[Function]}
                              onSubmit={[Function]}
                    />
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 18,
                                                  "id": 2,
                                                  "name": "VW",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
          </tbody>
</table>,
        "_debugID": 4,
        "_renderedOutput": <table>
          <thead>
                    <tr>
                              <th>
                                        Id
                              </th>
                              <th>
                                        Name
                              </th>
                              <th>
                                        acceleration
                              </th>
                    </tr>
          </thead>
          <tbody>
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 12,
                                                  "id": 1,
                                                  "name": "Audi",
                                                }
                              }
                              editEnabled={true}
                              onClick={[Function]}
                    />
                    <CarFormRow
                              car={
                                        Object {
                                                  "acceleration": 12,
                                                  "id": 1,
                                                  "name": "Audi",
                                                }
                              }
                              onCancel={[Function]}
                              onSubmit={[Function]}
                    />
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 18,
                                                  "id": 2,
                                                  "name": "VW",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
          </tbody>
</table>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CarTablePage
    cars={
        Array [
            Object {
              "acceleration": 12,
              "id": 1,
              "name": "Audi",
            },
            Object {
              "acceleration": 18,
              "id": 2,
              "name": "VW",
            },
          ]
    }
    editingCarId={1}
    onCancelCarChanges={[Function]}
    onRowClick={[Function]}
    onSaveCarChanges={[Function]}
/>,
}
`;

exports[`CarTablePage renders correctly with invalid editingCarId 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <table>
    <thead>
        <tr>
            <th>
                Id
            </th>
            <th>
                Name
            </th>
            <th>
                acceleration
            </th>
        </tr>
    </thead>
    <tbody>
        <CarRow
            car={
                Object {
                    "acceleration": 12,
                    "id": 1,
                    "name": "Audi",
                  }
            }
            editEnabled={false}
            onClick={[Function]}
        />
        <CarRow
            car={
                Object {
                    "acceleration": 18,
                    "id": 2,
                    "name": "VW",
                  }
            }
            editEnabled={false}
            onClick={[Function]}
        />
    </tbody>
</table>,
  "nodes": Array [
    <table>
      <thead>
            <tr>
                  <th>
                        Id
                  </th>
                  <th>
                        Name
                  </th>
                  <th>
                        acceleration
                  </th>
            </tr>
      </thead>
      <tbody>
            <CarRow
                  car={
                        Object {
                              "acceleration": 12,
                              "id": 1,
                              "name": "Audi",
                            }
                  }
                  editEnabled={false}
                  onClick={[Function]}
            />
            <CarRow
                  car={
                        Object {
                              "acceleration": 18,
                              "id": 2,
                              "name": "VW",
                            }
                  }
                  editEnabled={false}
                  onClick={[Function]}
            />
      </tbody>
</table>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <CarTablePage
        cars={
                Array [
                        Object {
                          "acceleration": 12,
                          "id": 1,
                          "name": "Audi",
                        },
                        Object {
                          "acceleration": 18,
                          "id": 2,
                          "name": "VW",
                        },
                      ]
        }
        editingCarId={100}
        onCancelCarChanges={[Function]}
        onRowClick={[Function]}
        onSaveCarChanges={[Function]}
/>,
      "_debugID": 7,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "cars": Array [
            Object {
              "acceleration": 12,
              "id": 1,
              "name": "Audi",
            },
            Object {
              "acceleration": 18,
              "id": 2,
              "name": "VW",
            },
          ],
          "editingCarId": 100,
          "onCancelCarChanges": [Function],
          "onRowClick": [Function],
          "onSaveCarChanges": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 4,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <table>
          <thead>
                    <tr>
                              <th>
                                        Id
                              </th>
                              <th>
                                        Name
                              </th>
                              <th>
                                        acceleration
                              </th>
                    </tr>
          </thead>
          <tbody>
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 12,
                                                  "id": 1,
                                                  "name": "Audi",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 18,
                                                  "id": 2,
                                                  "name": "VW",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
          </tbody>
</table>,
        "_debugID": 8,
        "_renderedOutput": <table>
          <thead>
                    <tr>
                              <th>
                                        Id
                              </th>
                              <th>
                                        Name
                              </th>
                              <th>
                                        acceleration
                              </th>
                    </tr>
          </thead>
          <tbody>
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 12,
                                                  "id": 1,
                                                  "name": "Audi",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 18,
                                                  "id": 2,
                                                  "name": "VW",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
          </tbody>
</table>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CarTablePage
    cars={
        Array [
            Object {
              "acceleration": 12,
              "id": 1,
              "name": "Audi",
            },
            Object {
              "acceleration": 18,
              "id": 2,
              "name": "VW",
            },
          ]
    }
    editingCarId={100}
    onCancelCarChanges={[Function]}
    onRowClick={[Function]}
    onSaveCarChanges={[Function]}
/>,
}
`;

exports[`CarTablePage renders correctly with last car in edit mode 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <table>
    <thead>
        <tr>
            <th>
                Id
            </th>
            <th>
                Name
            </th>
            <th>
                acceleration
            </th>
        </tr>
    </thead>
    <tbody>
        <CarRow
            car={
                Object {
                    "acceleration": 12,
                    "id": 1,
                    "name": "Audi",
                  }
            }
            editEnabled={false}
            onClick={[Function]}
        />
        <CarRow
            car={
                Object {
                    "acceleration": 18,
                    "id": 2,
                    "name": "VW",
                  }
            }
            editEnabled={true}
            onClick={[Function]}
        />
        <CarFormRow
            car={
                Object {
                    "acceleration": 18,
                    "id": 2,
                    "name": "VW",
                  }
            }
            onCancel={[Function]}
            onSubmit={[Function]}
        />
    </tbody>
</table>,
  "nodes": Array [
    <table>
      <thead>
            <tr>
                  <th>
                        Id
                  </th>
                  <th>
                        Name
                  </th>
                  <th>
                        acceleration
                  </th>
            </tr>
      </thead>
      <tbody>
            <CarRow
                  car={
                        Object {
                              "acceleration": 12,
                              "id": 1,
                              "name": "Audi",
                            }
                  }
                  editEnabled={false}
                  onClick={[Function]}
            />
            <CarRow
                  car={
                        Object {
                              "acceleration": 18,
                              "id": 2,
                              "name": "VW",
                            }
                  }
                  editEnabled={true}
                  onClick={[Function]}
            />
            <CarFormRow
                  car={
                        Object {
                              "acceleration": 18,
                              "id": 2,
                              "name": "VW",
                            }
                  }
                  onCancel={[Function]}
                  onSubmit={[Function]}
            />
      </tbody>
</table>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <CarTablePage
        cars={
                Array [
                        Object {
                          "acceleration": 12,
                          "id": 1,
                          "name": "Audi",
                        },
                        Object {
                          "acceleration": 18,
                          "id": 2,
                          "name": "VW",
                        },
                      ]
        }
        editingCarId={2}
        onCancelCarChanges={[Function]}
        onRowClick={[Function]}
        onSaveCarChanges={[Function]}
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "cars": Array [
            Object {
              "acceleration": 12,
              "id": 1,
              "name": "Audi",
            },
            Object {
              "acceleration": 18,
              "id": 2,
              "name": "VW",
            },
          ],
          "editingCarId": 2,
          "onCancelCarChanges": [Function],
          "onRowClick": [Function],
          "onSaveCarChanges": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <table>
          <thead>
                    <tr>
                              <th>
                                        Id
                              </th>
                              <th>
                                        Name
                              </th>
                              <th>
                                        acceleration
                              </th>
                    </tr>
          </thead>
          <tbody>
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 12,
                                                  "id": 1,
                                                  "name": "Audi",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 18,
                                                  "id": 2,
                                                  "name": "VW",
                                                }
                              }
                              editEnabled={true}
                              onClick={[Function]}
                    />
                    <CarFormRow
                              car={
                                        Object {
                                                  "acceleration": 18,
                                                  "id": 2,
                                                  "name": "VW",
                                                }
                              }
                              onCancel={[Function]}
                              onSubmit={[Function]}
                    />
          </tbody>
</table>,
        "_debugID": 6,
        "_renderedOutput": <table>
          <thead>
                    <tr>
                              <th>
                                        Id
                              </th>
                              <th>
                                        Name
                              </th>
                              <th>
                                        acceleration
                              </th>
                    </tr>
          </thead>
          <tbody>
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 12,
                                                  "id": 1,
                                                  "name": "Audi",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 18,
                                                  "id": 2,
                                                  "name": "VW",
                                                }
                              }
                              editEnabled={true}
                              onClick={[Function]}
                    />
                    <CarFormRow
                              car={
                                        Object {
                                                  "acceleration": 18,
                                                  "id": 2,
                                                  "name": "VW",
                                                }
                              }
                              onCancel={[Function]}
                              onSubmit={[Function]}
                    />
          </tbody>
</table>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CarTablePage
    cars={
        Array [
            Object {
              "acceleration": 12,
              "id": 1,
              "name": "Audi",
            },
            Object {
              "acceleration": 18,
              "id": 2,
              "name": "VW",
            },
          ]
    }
    editingCarId={2}
    onCancelCarChanges={[Function]}
    onRowClick={[Function]}
    onSaveCarChanges={[Function]}
/>,
}
`;

exports[`CarTablePage renders correctly with no edited car 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <table>
    <thead>
        <tr>
            <th>
                Id
            </th>
            <th>
                Name
            </th>
            <th>
                acceleration
            </th>
        </tr>
    </thead>
    <tbody>
        <CarRow
            car={
                Object {
                    "acceleration": 12,
                    "id": 1,
                    "name": "Audi",
                  }
            }
            editEnabled={false}
            onClick={[Function]}
        />
        <CarRow
            car={
                Object {
                    "acceleration": 18,
                    "id": 2,
                    "name": "VW",
                  }
            }
            editEnabled={false}
            onClick={[Function]}
        />
    </tbody>
</table>,
  "nodes": Array [
    <table>
      <thead>
            <tr>
                  <th>
                        Id
                  </th>
                  <th>
                        Name
                  </th>
                  <th>
                        acceleration
                  </th>
            </tr>
      </thead>
      <tbody>
            <CarRow
                  car={
                        Object {
                              "acceleration": 12,
                              "id": 1,
                              "name": "Audi",
                            }
                  }
                  editEnabled={false}
                  onClick={[Function]}
            />
            <CarRow
                  car={
                        Object {
                              "acceleration": 18,
                              "id": 2,
                              "name": "VW",
                            }
                  }
                  editEnabled={false}
                  onClick={[Function]}
            />
      </tbody>
</table>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <CarTablePage
        cars={
                Array [
                        Object {
                          "acceleration": 12,
                          "id": 1,
                          "name": "Audi",
                        },
                        Object {
                          "acceleration": 18,
                          "id": 2,
                          "name": "VW",
                        },
                      ]
        }
        editingCarId={-1}
        onCancelCarChanges={[Function]}
        onRowClick={[Function]}
        onSaveCarChanges={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "cars": Array [
            Object {
              "acceleration": 12,
              "id": 1,
              "name": "Audi",
            },
            Object {
              "acceleration": 18,
              "id": 2,
              "name": "VW",
            },
          ],
          "editingCarId": -1,
          "onCancelCarChanges": [Function],
          "onRowClick": [Function],
          "onSaveCarChanges": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <table>
          <thead>
                    <tr>
                              <th>
                                        Id
                              </th>
                              <th>
                                        Name
                              </th>
                              <th>
                                        acceleration
                              </th>
                    </tr>
          </thead>
          <tbody>
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 12,
                                                  "id": 1,
                                                  "name": "Audi",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 18,
                                                  "id": 2,
                                                  "name": "VW",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
          </tbody>
</table>,
        "_debugID": 2,
        "_renderedOutput": <table>
          <thead>
                    <tr>
                              <th>
                                        Id
                              </th>
                              <th>
                                        Name
                              </th>
                              <th>
                                        acceleration
                              </th>
                    </tr>
          </thead>
          <tbody>
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 12,
                                                  "id": 1,
                                                  "name": "Audi",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
                    <CarRow
                              car={
                                        Object {
                                                  "acceleration": 18,
                                                  "id": 2,
                                                  "name": "VW",
                                                }
                              }
                              editEnabled={false}
                              onClick={[Function]}
                    />
          </tbody>
</table>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CarTablePage
    cars={
        Array [
            Object {
              "acceleration": 12,
              "id": 1,
              "name": "Audi",
            },
            Object {
              "acceleration": 18,
              "id": 2,
              "name": "VW",
            },
          ]
    }
    editingCarId={-1}
    onCancelCarChanges={[Function]}
    onRowClick={[Function]}
    onSaveCarChanges={[Function]}
/>,
}
`;
